wal/watcher.go

```diff
@@ -0,0 +1,580 @@
+// Copyright 2018 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wal
+
+import (
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"path"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/tsdb/fileutil"
+	"github.com/prometheus/tsdb/record"
+)
+
+const (
+	readPeriod         = 10 * time.Millisecond
+	checkpointPeriod   = 5 * time.Second
+	segmentCheckPeriod = 100 * time.Millisecond
+	consumer           = "consumer"
+)
+
+// fromTime returns a new millisecond timestamp from a time.
+// This function is copied from prometheus/prometheus/pkg/timestamp to avoid adding vendor to TSDB repo.
+func fromTime(t time.Time) int64 {
+	return t.Unix()*1000 + int64(t.Nanosecond())/int64(time.Millisecond)
+}
+
+type WriteTo interface {
+	Append([]record.RefSample) bool
+	StoreSeries([]record.RefSeries, int)
+	SeriesReset(int)
+}
+
+type watcherMetrics struct {
+	recordsRead           *prometheus.CounterVec
+	recordDecodeFails     *prometheus.CounterVec
+	samplesSentPreTailing *prometheus.CounterVec
+	currentSegment        *prometheus.GaugeVec
+}
+
+// Watcher watches the TSDB WAL for a given WriteTo.
+type Watcher struct {
+	name           string
+	writer         WriteTo
+	logger         log.Logger
+	walDir         string
+	lastCheckpoint string
+	metrics        *watcherMetrics
+	readerMetrics  *liveReaderMetrics
+
+	startTime int64
+
+	recordsReadMetric       *prometheus.CounterVec
+	recordDecodeFailsMetric prometheus.Counter
+	samplesSentPreTailing   prometheus.Counter
+	currentSegmentMetric    prometheus.Gauge
+
+	quit chan struct{}
+	done chan struct{}
+
+	// For testing, stop when we hit this segment.
+	maxSegment int
+}
+
+func NewWatcherMetrics(reg prometheus.Registerer) *watcherMetrics {
```

Comment:
Should these take a consumer and provide an `Unregister` so that if a consumer disappears it can be removed from the metrics?

Reply:
We can't `Unregister` as all the Watcher metrics are under the same metric but just `WithLabels`. Both QueueManager and Watcher `Stop` delete their metrics via `DeleteLabelValues`.

Reply:
:+1: